plugins/out_kinesis_firehose/firehose_api.c

```diff
@@ -216,14 +217,26 @@ static int process_event(struct flb_firehose *ctx, struct flush *buf,
                          ctx->delivery_stream);
             return 2;
         }
-        /* guess space needed to write time_key */
-        len = 6 + strlen(ctx->time_key) + 6 * strlen(ctx->time_key_format);
+
+        /* format time output and return the length */
+        len = flb_aws_strftime_precision(&out_buf, ctx->time_key_format, tms);
+
         /* how much space do we have left */
         tmp_size = (buf->tmp_buf_size - buf->tmp_buf_offset) - written;
         if (len > tmp_size) {
-            /* not enough space- tell caller to retry */
+            /* not enough space - tell caller to retry */
+            flb_free(out_buf);
             return 1;
         }
+
+        if (len == -1) {
```

Comment:
nit: this failure case check happens after the length check above. I think it'd be most typical to check for error return first, and then check for a successful return with a time format that is too large.